ddtrace/contrib/pytest/_efd_utils.py

```diff
@@ -0,0 +1,243 @@
+        if call_report.outcome == _EFD_RETRY_OUTCOMES.EFD_ATTEMPT_FAILED:
+            _outcome_status = TestStatus.FAIL
+            if call_call.excinfo is not None:
+                _outcome_exc_info = TestExcInfo(call_call.excinfo.type, call_call.excinfo.value, call_call.excinfo.tb)
+                item.stash[caplog_records_key] = {}
+                item.stash[caplog_handler_key] = {}
+        elif call_report.skipped:
+            # A test that skips during retries is considered a failure
+            _outcome_status = TestStatus.FAIL
+        elif call_report.outcome == _EFD_RETRY_OUTCOMES.EFD_ATTEMPT_PASSED:
+            _outcome_status = TestStatus.PASS
+
+    # Teardown
+    teardown_call, teardown_report = _retry_run_when(item, "teardown", outcomes)
+    # Only override the outcome if the teardown failed, otherwise defer to either setup or call outcome
+    if teardown_report.outcome == _EFD_RETRY_OUTCOMES.EFD_ATTEMPT_FAILED:
+        _outcome_status = TestStatus.FAIL
+        if teardown_call.excinfo is not None:
+            _outcome_exc_info = TestExcInfo(
+                teardown_call.excinfo.type, teardown_call.excinfo.value, teardown_call.excinfo.tb
+            )
+            item.stash[caplog_records_key] = {}
+            item.stash[caplog_handler_key] = {}
+
+    item._initrequest()
+
+    return _TestOutcome(status=_outcome_status, skip_reason=_outcome_skip_reason, exc_info=_outcome_exc_info)
+
+
+def efd_pytest_terminal_summary(terminalreporter):
+    terminalreporter.write_sep("=", "Datadog Early Flake Detection", purple=True, bold=True)
+    # Print summary info
+    passed_reports = terminalreporter.getreports(_EFD_RETRY_OUTCOMES.EFD_FINAL_PASSED)
+    if passed_reports:
+        terminalreporter.write_sep("_", "PASSED", green=True, bold=True)
+        for passed_report in passed_reports:
+            line = f"{terminalreporter._tw.markup('PASSED', green=True)} {passed_report.nodeid}"
+            terminalreporter.write_line(line)
+        del terminalreporter.stats[_EFD_RETRY_OUTCOMES.EFD_FINAL_PASSED]
+    failed_reports = terminalreporter.getreports(_EFD_RETRY_OUTCOMES.EFD_FINAL_FAILED)
+    if failed_reports:
+        terminalreporter.write_sep("_", "FAILED", red=True, bold=True)
+        for failed_report in failed_reports:
+            line = f"{terminalreporter._tw.markup('FAILED', red=True)} {failed_report.nodeid}"
+            terminalreporter.write_line(line)
+        del terminalreporter.stats[_EFD_RETRY_OUTCOMES.EFD_FINAL_FAILED]
+    skipped_reports = terminalreporter.getreports(_EFD_RETRY_OUTCOMES.EFD_FINAL_SKIPPED)
+    if skipped_reports:
+        terminalreporter.write_sep("_", "SKIPPED", yellow=True, bold=True)
+        for skipped_report in skipped_reports:
+            line = f"{terminalreporter._tw.markup('SKIPPED', yellow=True)} {skipped_report.nodeid}"
+            terminalreporter.write_line(line)
+        del terminalreporter.stats[_EFD_RETRY_OUTCOMES.EFD_FINAL_SKIPPED]
```

Comment:
Why do we delete the dict entries here (but not for flaky reports below)?

Reply:
The stats determine what gets shown in the final line, so if we leave the other items in, they'll show up, eg:

```
==== 72 passed, 2 errors, 61 dd_efd_attempt_failed, 4 Flaky in 0.19s ====
```

This combines with the `initial_size` shenanigans in `pytest_terminal_summary` to try and make things work nicely.